File: src/test/kotlin/fr/unilim/saes5/unit/service/CompletionServiceTest.java
package fr.unilim.saes5.unit.service;

import fr.unilim.saes5.service.CompletionService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class CompletionServiceTest {

    private CompletionService completionService;

    @BeforeEach
    public void setUp() {
        this.completionService = new CompletionService();
    }

    @Test
    public void testAddCompletionWithNonNullValue() {
        this.completionService.addCompletion("hello");
        Set<String> completions = this.completionService.suggestCompletions("he");
        assertEquals(1, completions.size());
        assertTrue(completions.contains("hello"));
    }

    @Test
    public void testAddCompletionWithNullValue() {
        this.completionService.addCompletion(null);
        Set<String> completions = this.completionService.suggestCompletions("anything");
        assertTrue(completions.isEmpty());
    }

    @Test
    public void testAddCompletionWithTrimming() {
        this.completionService.addCompletion("  spaces  ");
        Set<String> completions = this.completionService.suggestCompletions("sp");
        assertEquals(1, completions.size());
        assertTrue(completions.contains("spaces"));
    }

    @Test
    public void testSuggestCompletionsWithNullInput() {
        this.completionService.addCompletion("test");
        Set<String> completions = this.completionService.suggestCompletions(null);
        assertTrue(completions.isEmpty());
    }

    @Test
    public void testSuggestCompletionsWithEmptyInput() {
        this.completionService.addCompletion("test");
        Set<String> completions = this.completionService.suggestCompletions("");
        assertEquals(1, completions.size());
    }

    @Test
    public void testSuggestCompletionsWithCaseInsensitivity() {
        this.completionService.addCompletion("CaseTest");
        Set<String> completions = this.completionService.suggestCompletions("caset");
        assertEquals(1, completions.size());
        assertTrue(completions.contains("CaseTest"));
    }

    @Test
    public void testSuggestCompletionsWithPartialMatch() {
        this.completionService.addCompletion("partial");
        Set<String> completions = this.completionService.suggestCompletions("art");
        assertEquals(1, completions.size());
        assertTrue(completions.contains("partial"));
    }

    @Test
    public void testSuggestCompletionsWithNoMatch() {
        this.completionService.addCompletion("nomatch");
        Set<String> completions = this.completionService.suggestCompletions("xyz");
        assertTrue(completions.isEmpty());
    }

    @Test
    public void testSuggestCompletionsWithMultipleMatches() {
        this.completionService.addCompletion("multi");
        this.completionService.addCompletion("multiplication");
        this.completionService.addCompletion("multiverse");
        this.completionService.addCompletion("museum");
        Set<String> completions = this.completionService.suggestCompletions("multi");
        assertEquals(3, completions.size());
        assertTrue(completions.containsAll(List.of("multi", "multiplication", "multiverse")));
    }

    @Test
    public void testSuggestCompletionsWithExactMatch() {
        this.completionService.addCompletion("exact");
        Set<String> completions = this.completionService.suggestCompletions("exact");
        assertEquals(1, completions.size());
        assertTrue(completions.contains("exact"));
    }

    @Test
    public void testAddCompletionWithDuplicateValues() {
        this.completionService.addCompletion("duplicate");
        this.completionService.addCompletion("duplicate");
        Set<String> completions = this.completionService.suggestCompletions("dup");
        assertEquals(1, completions.size());
    }

    @Test
    public void testSuggestCompletionsWithSpecialCharacters() {
        this.completionService.addCompletion("special@character!");
        Set<String> completions = this.completionService.suggestCompletions("@char");
        assertEquals(1, completions.size());
        assertTrue(completions.contains("special@character!"));
    }

    @Test
    public void testSuggestCompletionsWithNumericValues() {
        this.completionService.addCompletion("123numeric");
        Set<String> completions = this.completionService.suggestCompletions("123");
        assertEquals(1, completions.size());
        assertTrue(completions.contains("123numeric"));
    }

    @Test
    public void testSuggestCompletionsWithMixedCaseAndSpecialCharacters() {
        String mixedCaseCompletion = "Mix3dCase&Special";
        this.completionService.addCompletion(mixedCaseCompletion);
        Set<String> completions = this.completionService.suggestCompletions("mix3d");
        assertEquals(1, completions.size());
        assertTrue(completions.contains(mixedCaseCompletion));
    }

    @Test
    public void testSuggestCompletionsWithWhitespaceInput() {
        this.completionService.addCompletion("whitespace");
        Set<String> completions = this.completionService.suggestCompletions("white");
        assertEquals(1, completions.size());
        assertTrue(completions.contains("whitespace"));
    }

    @Test
    public void testSuggestCompletionsWithNoInputAndMultipleCompletions() {
        this.completionService.addCompletion("first");
        this.completionService.addCompletion("second");
        Set<String> completions = this.completionService.suggestCompletions("");
        assertEquals(2, completions.size());
        assertTrue(completions.containsAll(List.of("first", "second")));
    }

    @Test
    public void testSuggestCompletionsWithInputNotAtStart() {
        this.completionService.addCompletion("unmatched");
        Set<String> completions = this.completionService.suggestCompletions("matched");
        assertTrue(completions.contains("unmatched"));
    }

    @Test
    public void testAddEmptyStringAsCompletion() {
        this.completionService.addCompletion("");
        Set<String> completions = this.completionService.suggestCompletions("anything");
        assertTrue(completions.isEmpty());
    }

    @Test
    public void testSuggestCompletionsWhenNoCompletionsAdded() {
        Set<String> completions = this.completionService.suggestCompletions("test");
        assertTrue(completions.isEmpty());
    }

    @Test
    public void testSuggestCompletionsWithWhitespaceOnlyInput() {
        this.completionService.addCompletion("whitespace");
        Set<String> completions = this.completionService.suggestCompletions("   ");
        assertTrue(completions.isEmpty());
    }
}

File: src/test/kotlin/fr/unilim/codelinguo/unit/service/CompletionServiceTest.kt
package fr.unilim.codelinguo.unit.service

import fr.unilim.codelinguo.service.CompletionService
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class CompletionServiceTest {
    private var completionService: CompletionService? = null

    @BeforeEach
    fun setUp() {
        completionService = CompletionService()
    }

    @Test
    fun testAddCompletionWithNonNullValue() {
        completionService!!.addCompletion("hello")
        val completions = completionService!!.suggestCompletions("he")
        Assertions.assertEquals(1, completions.size)
        Assertions.assertTrue(completions.contains("hello"))
    }

    @Test
    fun testAddCompletionWithNullValue() {
        completionService!!.addCompletion(null)
        val completions = completionService!!.suggestCompletions("anything")
        Assertions.assertTrue(completions.isEmpty())
    }

    @Test
    fun testAddCompletionWithTrimming() {
        completionService!!.addCompletion("  spaces  ")
        val completions = completionService!!.suggestCompletions("sp")
        Assertions.assertEquals(1, completions.size)
        Assertions.assertTrue(completions.contains("spaces"))
    }

    @Test
    fun testSuggestCompletionsWithNullInput() {
        completionService!!.addCompletion("test")
        val completions = completionService!!.suggestCompletions(null)
        Assertions.assertTrue(completions.isEmpty())
    }

    @Test
    fun testSuggestCompletionsWithEmptyInput() {
        completionService!!.addCompletion("test")
        val completions = completionService!!.suggestCompletions("")
        Assertions.assertEquals(1, completions.size)
    }

    @Test
    fun testSuggestCompletionsWithCaseInsensitivity() {
        completionService!!.addCompletion("CaseTest")
        val completions = completionService!!.suggestCompletions("caset")
        Assertions.assertEquals(1, completions.size)
        Assertions.assertTrue(completions.contains("CaseTest"))
    }

    @Test
    fun testSuggestCompletionsWithPartialMatch() {
        completionService!!.addCompletion("partial")
        val completions = completionService!!.suggestCompletions("art")
        Assertions.assertEquals(1, completions.size)
        Assertions.assertTrue(completions.contains("partial"))
    }

    @Test
    fun testSuggestCompletionsWithNoMatch() {
        completionService!!.addCompletion("nomatch")
        val completions = completionService!!.suggestCompletions("xyz")
        Assertions.assertTrue(completions.isEmpty())
    }

    @Test
    fun testSuggestCompletionsWithMultipleMatches() {
        completionService!!.addCompletion("multi")
        completionService!!.addCompletion("division et multiplication")
        completionService!!.addCompletion("le multiverse")
        completionService!!.addCompletion("musée")
        val completions = completionService!!.suggestCompletions("multi")
        Assertions.assertEquals(3, completions.size)
        Assertions.assertTrue(completions.containsAll(listOf("multi", "division et multiplication", "le multiverse")))
    }

    @Test
    fun testSuggestCompletionsWithExactMatch() {
        completionService!!.addCompletion("exact")
        val completions = completionService!!.suggestCompletions("exact")
        Assertions.assertEquals(1, completions.size)
        Assertions.assertTrue(completions.contains("exact"))
    }

    @Test
    fun testAddCompletionWithDuplicateValues() {
        completionService!!.addCompletion("duplicate")
        completionService!!.addCompletion("duplicate")
        val completions = completionService!!.suggestCompletions("dup")
        Assertions.assertEquals(1, completions.size)
    }

    @Test
    fun testSuggestCompletionsWithSpecialCharacters() {
        completionService!!.addCompletion("special@character!")
        val completions = completionService!!.suggestCompletions("@char")
        Assertions.assertEquals(1, completions.size)
        Assertions.assertTrue(completions.contains("special@character!"))
    }

    @Test
    fun testSuggestCompletionsWithNumericValues() {
        completionService!!.addCompletion("123numeric")
        val completions = completionService!!.suggestCompletions("123")
        Assertions.assertEquals(1, completions.size)
        Assertions.assertTrue(completions.contains("123numeric"))
    }

    @Test
    fun testSuggestCompletionsWithMixedCaseAndSpecialCharacters() {
        val mixedCaseCompletion = "Mix3dCase&Special"
        completionService!!.addCompletion(mixedCaseCompletion)
        val completions = completionService!!.suggestCompletions("mix3d")
        Assertions.assertEquals(1, completions.size)
        Assertions.assertTrue(completions.contains(mixedCaseCompletion))
    }

    @Test
    fun testSuggestCompletionsWithWhitespaceInput() {
        completionService!!.addCompletion("whitespace  ")
        val completions = completionService!!.suggestCompletions("white")
        Assertions.assertEquals(1, completions.size)
        Assertions.assertTrue(completions.contains("whitespace"))
    }

    @Test
    fun testSuggestCompletionsWithNoInputAndMultipleCompletions() {
        completionService!!.addCompletion("first")
        completionService!!.addCompletion("second")
        val completions = completionService!!.suggestCompletions("")
        Assertions.assertEquals(2, completions.size)
        Assertions.assertTrue(completions.containsAll(listOf("first", "second")))
    }

    @Test
    fun testSuggestCompletionsWithInputNotAtStart() {
        completionService!!.addCompletion("unmatched")
        val completions = completionService!!.suggestCompletions("matched")
        Assertions.assertTrue(completions.contains("unmatched"))
    }

    @Test
    fun testAddEmptyStringAsCompletion() {
        completionService!!.addCompletion("")
        val completions = completionService!!.suggestCompletions("anything")
        Assertions.assertTrue(completions.isEmpty())
    }

    @Test
    fun testSuggestCompletionsWhenNoCompletionsAdded() {
        val completions = completionService!!.suggestCompletions("test")
        Assertions.assertTrue(completions.isEmpty())
    }

    @Test
    fun testSuggestCompletionsWithWhitespaceOnlyInput() {
        completionService!!.addCompletion("whitespace")
        val completions = completionService!!.suggestCompletions("   ")
        Assertions.assertTrue(completions.isEmpty())
    }
}
